### PR TITLE
Refactor the `wc` command to organize the counts

### PR DESCRIPTION
**Pull request by stratact**

### CODE REVIEW

File: src/bin/wc.rs [hunk 1]
```diff
@@ -39,7 +41,63 @@ static MAN_PAGE: &'static str = r#"
...
+#[derive(Copy, Clone, Default)]
+struct Counter {
+    lines: u64,
+    words: u64,
+    bytes: u64,
+}
+
+impl Counter {
+    #[inline]
+    fn new<T: std::io::Read>(input: T) -> Self {
```
Comment (Ticki) [inline src/bin/wc.rs]: Why not import the `Read` trait instead?

File: src/bin/wc.rs [hunk 2]
```diff
@@ -47,27 +105,19 @@ struct Flags {
...
-        }
-    }
-
-    fn print_count<'a, W: Write>(self, lines: u64, words: u64, bytes: u64, file: &'a str, stdout: &mut W, stderr: &mut Stderr) {
+    fn print_count<'a, W: Write>(self, count: Counter, file: &'a str, stdout: &mut W, stderr: &mut Stderr) {
         stdout.write(b"    ").try(&mut *stderr);
 
         if self.count_lines {
-            stdout.write(lines.to_string().as_bytes()).try(&mut *stderr);
+            stdout.write(count.lines.to_string().as_bytes()).try(&mut *stderr);
```
Comment (Ticki) [inline src/bin/wc.rs]: `to_owned`, `to_owned`, `to_ooooowned` :notes:

File: src/bin/wc.rs [hunk 3]
```diff
@@ -47,27 +105,19 @@ struct Flags {
...
             stdout.write(b" ").try(&mut *stderr);
         }
         if self.count_words {
-            stdout.write(words.to_string().as_bytes()).try(&mut *stderr);
+            stdout.write(count.words.to_string().as_bytes()).try(&mut *stderr);
```
Comment (Ticki) [inline src/bin/wc.rs]: Same here.

File: src/bin/wc.rs [hunk 4]
```diff
@@ -47,27 +105,19 @@ struct Flags {
...
             stdout.write(b" ").try(&mut *stderr);
         }
         if self.count_bytes {
-            stdout.write(bytes.to_string().as_bytes()).try(&mut *stderr);
+            stdout.write(count.bytes.to_string().as_bytes()).try(&mut *stderr);
```
Comment (Ticki) [inline src/bin/wc.rs]: And here.

File: src/bin/wc.rs (hunk 1 above)
Comment (stratact) [inline src/bin/wc.rs]: I don't see why not? I just move over the `do_count` function to be the constructor.

Comment (Ticki) [inline src/bin/wc.rs]: I mean: on line 53, you should replace `std::io::Read` with `Read`.

**Comment by Ticki** [comment]
LFTM!

**Comment by Ticki** [comment]
You might want to rebase, there are double-commits.

**Comment by Ticki** [comment]
When that's done, I'll merge it.

**Comment by stratact** [comment]
Refresh the browser

**Comment by Ticki** [comment]
Thanks!

**Pull request merged**